org.eclipse.lsp4e/src/org/eclipse/lsp4e/internal/FutureUtil.java

```diff
@@ -20,25 +22,20 @@
 public class FutureUtil {
 
 	/**
-	 * Combines two async lists of results into a single list by adding all the
-	 * elements of the second list to the first one.
+	 * Combines two async lists of results into a new single list containing all
+	 * elements of the first and the second result.
 	 *
-	 * @param <T>
-	 *            Result type
-	 * @param accumulator
-	 *            One async result
-	 * @param another
-	 *            Another async result
 	 * @return Async combined result
 	 */
-	public static <T> CompletableFuture<List<T>> addAll(CompletableFuture<List<T>> accumulator,
-			CompletableFuture<List<T>> another) {
-		CompletableFuture<List<T>> res = accumulator.thenCombine(another, (a, b) -> {
-			a.addAll(b);
-			return a;
+	public static <T> CompletableFuture<List<T>> join(final CompletableFuture<List<T>> firstResult,
+			final CompletableFuture<List<T>> secondResult) {
+		final CompletableFuture<List<T>> result = firstResult.thenCombine(secondResult, (firstList, secondList) -> {
+			final List<T> combinedList = new ArrayList<>(firstList);
```

Comment:
I am unsure about this. The new implementation is safer, but it is also going to make unnecessary copies in many cases.

Could we not check before creating a new list if the old one is modifiable?

I know it is a bit ugly but it is probably worth to make a copy only if the list is a Guava ImmutableList or an instance of java.util.ImmutableCollections?

Otherwise I think I would be happier assuming the lists used for the join are not immutable.

Reply:
I think it is any case not good practice to modify the list in-place that is held by a future. If the first future is for example accessed through another code path too it suddenly contains elements from unrelated futures. I think the way `addAll` is implemented it is so specific and with side effects that it should not be exposed at all and only privately declared and inside where needed.